command/agent/dns.go

```diff
@@ -33,6 +33,37 @@ type DNSServer struct {
 	logger       *log.Logger
 }
 
+// Shutdown stops the DNS Servers
+func (d *DNSServer) Shutdown() error {
+	// Async stop the DNS Servers, handle a potential error
+	errCh := make(chan error, 1)
+	go func(d *DNSServer) {
+		if err := d.dnsServer.Shutdown(); err != nil {
+			d.logger.Printf("[ERR] dns: error stopping udp server: %v", err)
+			errCh <- fmt.Errorf("dns udp teardown failed: %v", err)
+		}
+	}(d)
+
+	errChTCP := make(chan error, 1)
+	go func(d *DNSServer) {
+		if err := d.dnsServerTCP.Shutdown(); err != nil {
+			d.logger.Printf("[ERR] dns: error stopping tcp server: %v", err)
+			errChTCP <- fmt.Errorf("dns tcp teardown failed: %v", err)
+		}
+	}(d)
+
+	// Wait for either the check, listen error, or timeout
```

Comment:
Ooops - Forgot about that third signal. Shutdown only takes a few `ms` but if there are lookups in progress this could theoretically be longer [see docs](https://github.com/miekg/dns/blob/2b6f0d61e56de5436d7b964e3f667b76d34adb0f/server.go#L315). Doing the shutdown in sequence would be much simpler - if you prefer this, give me a :+1: and I'll update the PR